src/gui/guiFormSpecMenu.cpp

```diff
@@ -513,14 +514,30 @@ void GUIFormSpecMenu::parseScrollBar(parserData* data, const std::string &elemen
 
 		spec.ftype = f_ScrollBar;
 		spec.send  = true;
-		gui::IGUIScrollBar* e =
-				Environment->addScrollBar(is_horizontal,rect,this,spec.fid);
+		guiScrollBar* e = new guiScrollBar(Environment, this, spec.fid, rect,
```

Comment:
These references must be 'dropped' in the formspec destructor otherwise it will lead to memory leaks!

Reply:
[Destructer](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L119) and [regenerateGui](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L2057) call [`removeChildren`](https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L159) [which drops all children](http://irrlicht.sourceforge.net/docu/_i_g_u_i_element_8h_source.html#l00304) of `this` which include `e`.

Reply:
No, the children are removed, not dropped. Elements that are 'added' with `Evironment-add..` should not be dropped.

Reply:
Oh, [`addChildToEnd`](http://irrlicht.sourceforge.net/docu/_i_g_u_i_element_8h_source.html#l00818) grabs the element. Well, then I think, it would be ok to drop `e` directly after constructing it.

Reply:
I suggest you look at how tables are handled `m_tables` there are also created with the `new` operator, hence why they are being 'dropped' here: https://github.com/minetest/minetest/blob/5106ea9e39a4ac34e645baae419a5ad1f2a5cb91/src/gui/guiFormSpecMenu.cpp#L122

Essentially, the author needs to do the same for `m_scrollbars`. It may be possible to drop them before hand, however, it is more usual to do this in the destructor.

Reply:
Maybe just drop the element directly after creating it if it is never used again directly by this class.

Reply:
Are you sure, or is this a guess? I would prefer to err on the side of caution and consistency and do it like they did it for tables :)

Reply:
I'm not really a trustworthy source here. Other humans are more experienced.

Reply:
I think I have this done.  I just copied the table and changed it a bit, so I want someone to check it over.

Reply:
It looks good now.